### PR TITLE
[MERGE] web,project: add burndown chart in reporting menu

### PR DESCRIPTION
**Pull request by xavierbol**
Purpose
-----------

Currently, we have reports that depict the status of a
project at a certain point in time, but we don't have any reports
representing the actual progress of the project in time.

### Burndown Chart

A burndown chart is added in the reporting menu, this chart would help users see
the evolution of the project and determine whether it is on the
right track or not.

### Stacked in line chart

Currently, we can stacked in bar chart but not in line and pie charts. Here, the stacked will be also available in the line chart for the burndown chart.
Some unit tests in JS are added to keep the stacked in line chart of the burndown chart.

### Disable the pie chart in this reporting

This chart is not useful for this reporting, thus, the button has been hidden.

### Move constants and functions in own JS file

In web module, the GraphRenderer file contains some constants and functions used in the renderer of the graph view, these constants and functions are moved in new file called graph_utils.js to easily export these.

task-2458017

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/project/static/tests/burndown_chart_tests.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+            burndown_chart.destroy();
+        });
+
+        QUnit.test('check if the line chart is staked when the staked props is True', async function (assert) {
+            assert.expect(5);
+
+            const burndown_chart = await createView(this.burndown_chart);
+            assert.ok(burndown_chart.renderer.props.stacked, "should be stacked by default.");
+            assert.strictEqual(burndown_chart.renderer.props.mode, "line", "should be in line chart mode by default.");
```
Comment (IT-Ideas) [inline addons/project/static/tests/burndown_chart_tests.js:92]: Should we retest this as it was tested in the previous test?

File: addons/project/static/tests/burndown_chart_tests.js [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            burndown_chart.destroy();
+        });
+
+        QUnit.test('check if it is classic line chart when stacked prop is false', async function (assert) {
+            assert.expect(6);
+
+            const burndown_chart = await createView(this.burndown_chart);
+
+            assert.ok(burndown_chart.renderer.props.stacked, "should be stacked by default.");
+            assert.strictEqual(burndown_chart.renderer.props.mode, "line", "should be in line chart mode by default.");
```
Comment (IT-Ideas) [inline addons/project/static/tests/burndown_chart_tests.js:132]: Should we retest this as it was tested in the previous test?

**Review by IT-Ideas — COMMENTED** [review]
Great Job 🙂 
Just a few small questions.

Thanks for your work!

File: addons/project/report/project_task_burndown_chart_report_views.xml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                <filter string="Unassigned" name="unassigned" domain="[('user_id', '=', False)]"/>
+                <separator/>
+                <filter name="filter_date" date="date" string="Date" />
+                <filter name="filter_date_deadline" date="date_deadline"/>
+                <filter name="filter_date_assign" date="date_assign"/>
+                <group expand="0" string="Group By">
+                    <filter string="Date" name="date" context="{'group_by': 'date'}" />
+                    <filter string="Stage" name="stage_id" context="{'group_by': 'stage_id'}" />
+                    <filter string="Project" name="project_id" context="{'group_by': 'project_id'}" />
+                    <filter string="Assigne To" name="user_id" context="{'group_by': 'user_id'}" />
```
[the remark this answers is not in the document]
Comment (xavierbol) [reply addons/project/report/project_task_burndown_chart_report_views.xml]: Oh, indeed, thanks !

File: addons/project/static/tests/burndown_chart_tests.js (hunk 1 above)
Comment (xavierbol) [reply addons/project/static/tests/burndown_chart_tests.js:92]: I do this to be sure that stacked and mode are the same in each test, for instance, if someone changes the default view for the burndown chart, he will must check all tests and not only a test with these both lines. 
However, it is true the next assertions can be false if the mode is not line chart and stacked is not true. Thus, I'm agree with you, we can only do it in the first test. If you prefer to test it only once, I will change it.

Comment (IT-Ideas) [reply addons/project/static/tests/burndown_chart_tests.js:92]: The idea behind my question is to try to be the most atomic we can in the tests. But is you think it is required, no problem for me 🙂

Comment (xavierbol) [reply addons/project/static/tests/burndown_chart_tests.js]: Yeah, you are right, I changed it. 🙂

**Comment by tivisse** [comment]
@robodoo r+ merge

**Pull request merged**